File: src/lib/functions/amazonSearchTools.ts

import { FunctionDeclaration, SchemaType } from "@google/generative-ai";

export const searchAmazonProductTool: FunctionDeclaration = {
  name: "searchAmazonProduct",
  description: "Search for a specific product on Amazon using RapidAPI. Returns top 3 relevant products with details like title, price, rating, and link.",
  parameters: {
    type: SchemaType.OBJECT,
    properties: {
      product_query: {
        type: SchemaType.STRING,
        description: "The product name or search query to find on Amazon (e.g., 'organic bananas', 'whole wheat bread')",
      },
      country: {
        type: SchemaType.STRING,
        description: "Country code for Amazon search (default: 'US')",
      },
    },
    required: ["product_query"],
  },
};

export const searchMultipleAmazonProductsTool: FunctionDeclaration = {
  name: "searchMultipleAmazonProducts",
  description: "Search for multiple products from the shopping list on Amazon in a batch operation. This is efficient for finding all shopping list items at once.",
  parameters: {
    type: SchemaType.OBJECT,
    properties: {
      product_queries: {
        type: SchemaType.ARRAY,
        description: "Array of product names from the shopping list to search for on Amazon",
        items: {
          type: SchemaType.STRING,
        },
      },
      country: {
        type: SchemaType.STRING,
        description: "Country code for Amazon search (default: 'US')",
      },
    },
    required: ["product_queries"],
  },
};

export const getAmazonSearchResultsTool: FunctionDeclaration = {
  name: "getAmazonSearchResults",
  description: "Retrieve previously cached Amazon search results for shopping list items. Use this to show results without making new API calls.",
  parameters: {
    type: SchemaType.OBJECT,
    properties: {
      product_name: {
        type: SchemaType.STRING,
        description: "The product name to get cached results for (optional - if not provided, returns all cached results)",
      },
    },
  },
};

export const clearAmazonSearchCacheTool: FunctionDeclaration = {
  name: "clearAmazonSearchCache",
  description: "Clear all cached Amazon search results. Use when shopping list changes significantly.",
  parameters: {
    type: SchemaType.OBJECT,
    properties: {},
  },
};

File: mise-asi/registry/schemas/amazon_search_tools.py
"""
Amazon search tool schemas
Maps to: src/lib/functions/amazonSearchTools.ts
"""

search_amazon_product_tool = {
    "name": "searchAmazonProduct",
    "description": "Search for a product on Amazon. Returns top results with title, price, rating, and link.",
    "input_schema": {
        "type": "object",
        "properties": {
            "product_query": {"type": "string", "description": "Product search query"},
            "country": {"type": "string", "description": "Country code (default: US)", "default": "US"}
        },
        "required": ["product_query"]
    }
}

search_multiple_amazon_products_tool = {
    "name": "searchMultipleAmazonProducts",
    "description": "Search for multiple products on Amazon at once.",
    "input_schema": {
        "type": "object",
        "properties": {
            "queries": {
                "type": "array",
                "items": {"type": "string"},
                "description": "List of product search queries"
            },
            "country": {"type": "string", "default": "US"}
        },
        "required": ["queries"]
    }
}

get_amazon_search_results_tool = {
    "name": "getAmazonSearchResults",
    "description": "Get cached Amazon search results for the user.",
    "input_schema": {
        "type": "object",
        "properties": {},
        "required": []
    }
}

clear_amazon_search_cache_tool = {
    "name": "clearAmazonSearchCache",
    "description": "Clear the cached Amazon search results.",
    "input_schema": {
        "type": "object",
        "properties": {},
        "required": []
    }
}
